src/wallet/spend.cpp

```diff
@@ -615,15 +625,59 @@ static void DiscourageFeeSniping(CMutableTransaction& tx, FastRandomContext& rng
     // enough, that fee sniping isn't a problem yet, but by implementing a fix
     // now we ensure code won't be written that makes assumptions about
     // nLockTime that preclude a fix later.
+    //
+    // nSequence-based
+    // ---------------
+    //
+    // If all inputs in the tx spend taproot coins, use nSequence to discourage
+    // fee-sniping. This creates a "privacy cloak" for txs that use taproot
+    // key-path spends (BIP326).
+    //
+    // Compared to nLockTime-based anti-fee-sniping, this may be minimally less
+    // effective, as nSequence only allows relative locks. If all inputs of the
+    // tx can be moved backward in the chain, this tx can be as well. Though,
+    // this should be fine given that 10% of txs have their locktime randomly
+    // reduced to a past block. Also, sequence-based locking will only be used
+    // if every sequence is at least 1.
     if (IsCurrentForAntiFeeSniping(chain, block_hash)) {
-        tx.nLockTime = block_height;
-
-        // Secondly occasionally randomly pick a nLockTime even further back, so
-        // that transactions that are delayed after signing for whatever reason,
-        // e.g. high-latency mix networks and some CoinJoin implementations, have
-        // better privacy.
-        if (rng_fast.randrange(10) == 0) {
-            tx.nLockTime = std::max(0, int(tx.nLockTime) - int(rng_fast.randrange(100)));
+        auto locktime_based{[&] {
+            if (!SignalsOptInRBF(CTransaction{tx})) {
+                // Use locktime if the tx inputs do not opt-in to RBF
+                return true;
+            }
+            for (const auto& in : inputs) {
+                if (in.depth < 1 || uint64_t(in.depth) > MAX_BLOCK_SEQ) {
+                    // Use locktime if any depth is out-of-range
+                    return true;
+                }
+                std::vector<std::vector<uint8_t>> dummy;
+                const TxoutType in_type{Solver(in.txout.scriptPubKey, dummy)};
+                // Use locktime if any input isn't taproot
+                if (in_type != TxoutType::WITNESS_V1_TAPROOT) {
+                    return true;
+                }
```

Comment:
It has a specific usecase for Taproot, but what is the rationale for disabling it for non-(pure-)Taproot transactions? The "created by Bitcoin Core wallet" fingerprint?

Reply:
I can't say for sure, since I am not the author of BIP 326, but I presume:

* Any protocol that relies on sequence values in "legacy" key-only spends may be encouraged to upgrade to taproot.
* Any new "population" of txs that have the same fingerprint will reduce the privacy for themselves and everyone else. Though, if such a population already exists, I can see the reason for Bitcoin Core to mix itself into that population.

So I guess the question is whether such a protocol exists and is unlikely to use/upgrade to taproot?

Reply:
As the author of bip326, yes, I was thinking along the lines of @MarcoFalke's second point. Changing the behaviour of legacy transactions of which there is already a large population would reduce privacy. It's best to leave legacy transactions unchanged. Also non-taproot outputs cannot (easily) be used to create point-time-locked-contracts so the "privacy cloak" they create would never be used.

Reply:
I think a protocol that uses plain non-taproot keys with just plain pre-signed transactions and sequence based locktimes could still benefit from Bitcoin Core cloaking non-taproot inputs, no? Though, as they only use plain keys they might as well upgrade to taproot to get the taproot-specific cloak.

Reply:
I don't think that kind of protocol has a point of existing. Why would someone use nsequence-based locktimes when they can't have some kind of point-lock? A point lock requires taproot.

Reply:
Forgot to reply to this comment earlier. Chris, nSequence-based locktimes are useful for anti-fee sniping for presigned transactions (as you can't use nLockTime-based protection for those, use an input with an nSequence set when bumping the fees).
I don't think this is a justification for cloaking non-Taproot inputs here. By the time the cloak is realistically available, they would probably use Taproot and otherwise they should. They have an incentive to upgrade anyways, as fee-bumping with Taproot inputs is more efficient.